### PR TITLE
API-3.1: Add OAuth scopes

### PR DESCRIPTION
**Pull request by wvengen**
Part one of #572 (which was part of #429), continues from #577 (only the last commit here is relevant).

### CODE REVIEW

File: app/controllers/api/v1/base_controller.rb [hunk 1]
```diff
@@ -11,24 +13,16 @@ class Api::V1::BaseController < ApplicationController
...
   # @return [Ordergroup] Current user's ordergroup, or +nil+ if no valid token or user has no ordergroup.
   def current_ordergroup
     current_user.try(:ordergroup)
   end
 
   def require_ordergroup
     authenticate
-    raise Api::Errors::PermissionRequired unless current_user.ordergroup.present?
+    unless current_ordergroup.present?
+      raise Api::Errors::PermissionRequired.new("Forbidden, must be in an ordergroup")
```
Comment (paroga) [inline app/controllers/api/v1/base_controller.rb]: single quotes?

File: app/controllers/api/v1/base_controller.rb [hunk 2]
```diff
@@ -42,13 +36,18 @@ def not_found_handler(e)
   end
 
   def not_acceptable_handler(e)
-    render status: 422, json: {error: 'not_acceptable', error_description: e.message || 'Data not acceptable' }
+    msg = e.message || 'Data not acceptable'
+    render status: 422, json: {error: 'not_acceptable', error_description: msg }
```
Comment (paroga) [inline app/controllers/api/v1/base_controller.rb]: space before the closing bracket?

File: config/initializers/doorkeeper.rb [hunk 3]
```diff
@@ -47,8 +47,13 @@
...
   # https://github.com/doorkeeper-gem/doorkeeper/wiki/Using-Scopes
-  # default_scopes  :public
-  # optional_scopes :write, :update
+
+  # default is a collection of read-only scopes
+  default_scopes 'config:user', 'user:read'
+
+  optional_scopes 'config:read', 'config:write',
+                  'user:write',
+                  'offline_access'
```
Comment (paroga) [inline config/initializers/doorkeeper.rb]: do we use `offline_access` already somewhere?

File: app/controllers/concerns/auth_api.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    doorkeeper_authorize_roles!(*scopes) if scopes.present? && !performed?
+  end
+
+  private
+
+  # Make sure that at least one the given OAuth scopes is valid for the current user's permissions.
+  # @raise Api::Errors::PermissionsRequired
+  def doorkeeper_authorize_roles!(*scopes)
+    unless scopes.any? {|scope| doorkeeper_scope_permitted?(scope) }
+      raise Api::Errors::PermissionRequired.new("Forbidden, no permission")
```
Comment (paroga) [inline app/controllers/concerns/auth_api.rb]: singel quotes?

**Comment by wvengen** [comment]
Thanks for the review. The scopes don't feel too nice for me, but I don't see how to really improve that. I agree about being able to change them, until we do a release. I'll add a note to the API docs about that.
I guess this is also a green light for #577 (which this PR builds on).

File: config/initializers/doorkeeper.rb (hunk 3 above)
Comment (wvengen) [reply config/initializers/doorkeeper.rb]: It is used [here](https://github.com/foodcoops/foodsoft/blob/master/app/controllers/concerns/auth.rb#L138) (already on master).

**Pull request merged**